Add tag support

clearydude opened the pull request:
This PR adds support for secure-push/fetch of tags to git-rsl.

Closes #48

Commit:
Rename BranchName type

Prior to this commit, we were wrapping strings representing branch names into BranchName types to reduce potential for misuse.
This commit renames that to ReferenceName to reflect upcoming tag support.

Commit:
Add tag support for git push operation and test

Prior to this commit, the git methods for push were bound to using branches. This commit removes that and generalizes it to enable pushing a branch or tag ref. It adds a new test for tag pushing as well as a helper method to tag the latest commit in a repo. It also updates the integration tests to use the ReferenceName type.

Commit:
Change bin docs to REFERENCE

Prior to this commit, the help prompts for the binaries still told the user that they need to supply BRANCH references. This commit generalizes that to include branches OR tags by using REFERENCE

Commit:
Use reference name to find target

Prior to this commit, push entries would try to locate a branch using the shorthand name provided from the user. This was problematic because if the user was attempting to push a tag, the operation would fail. This commit changes it to instead use the git helper method to find a git2::Reference using the shorthand name, so that we can stash the target git2::Oid of either a branch OR a tag in the RSL push entry.

Commit:
Generalize finding remote push entries

Prior to this commit, the `find_remote_push_entry_for_branch` method tried to use the shorthand name to locate the full reference name. This commit changes it to instead find the reference, either a branch or a tag, and get that name instead. This allows this method to locate RSL push entries for tags as well as branches.

Commit:
Change some naming

Prior to this commit, methods to locate refererences that had been previously changed from locating branches had not been renamed to reflect their new functionality. This commit does all that stuff.

windwardrail commented on src/bin/git-secure-fetch.rs:
```diff
@@ -32,8 +32,8 @@ fn main() {
+        None => panic!("Must supply a REFERENCE argument"),
         Some(v) => v.to_owned(),
     };
     // TODO - reduce code duplication across the top level of the binaries
```
Move todo into an issue

clearydude replied:
9acbc9f and issue incoming

mullr commented on src/utils/git.rs:
```diff
@@ -267,21 +267,31 @@ pub fn fetch(
+fn get_ref_from_name<'repo>(repo: &'repo Repository, name: &str) -> Option<Reference<'repo>> {
+    match repo.find_branch(name, BranchType::Local) {
+        Ok(branch) => Some(branch.into_reference()),
+        Err(e) => { // toDo may be other errors here...
```
leftover todo

clearydude replied:
2fc4cdc

mullr commented on src/utils/git.rs:
```diff
@@ -267,21 +267,31 @@ pub fn fetch(
+    match repo.find_branch(name, BranchType::Local) {
+        Ok(branch) => Some(branch.into_reference()),
+        Err(e) => { // toDo may be other errors here...
+            let tag_ref = repo.find_reference(&format!("refs/tags/{}", name)).expect("reference not found");
```
`panic`ing when the reference isn't found definitely isn't what we should do, especially considering this can be used as a library.

clearydude replied:
638721c

mullr commented on src/utils/git.rs:
```diff
@@ -267,21 +267,31 @@ pub fn fetch(
+        let reference = get_ref_from_name(&repo, name).expect("failed to find reference from name");
+        let full_name = String::from(reference.name().expect("failed to get full name of ref"));
+        full_names.push(full_name)
+    }
```
this is more idiomatically done with `ref_names.map(|n| ...)`

clearydude replied:
b776d8f

mullr commented on src/utils/git.rs:
```diff
@@ -267,21 +267,31 @@ pub fn fetch(
+    let mut full_names: Vec<String> = vec![];
+    for name in ref_names {
+        let reference = get_ref_from_name(&repo, name).expect("failed to find reference from name");
+        let full_name = String::from(reference.name().expect("failed to get full name of ref"));
```
We shouldn't `panic` here either; basically, down in the internal modules, nothing should ever panic. It's more forgivable at the top level of a CLI application.

clearydude replied:
638721c

mullr commented on src/utils/test_helper.rs:
```diff
@@ -87,6 +87,13 @@ pub fn create_file_with_text<P: AsRef<Path>>(path: P, text: &str) -> () {
 }
 
+
+pub fn tag_lightweight<'repo>(repo: &'repo Repository, tag_name: &str) -> Reference<'repo> {
```
what does lightweight mean?

clearydude replied:
It's basically just a pointer vs an annotated tag with all kinds of other information. I feel like for the test I probably don't need to create a full on annotated tag, but am open to doing that instead if it makes more sense.

mullr replied:
If we already have signing set up for the integration tests we may as well?

clearydude replied:
Ok, I accept this

clearydude replied:
f687fb8

mullr commented on tests/integration_test.rs:
```diff
@@ -30,36 +30,54 @@ macro_rules! sequential_test {
+            do_work_on_branch(&context.local, "refs/heads/master");
+            // let local_tag = tag_lightweight(&mut context.local, "v6.66");
+
+            // git_rsl::utils::git::push(&context.local, &mut context.local.find_remote("origin").expect("failed to find remote"), &["v6.66"]);
```
should delete the commented lines

clearydude replied:
They're out as of 2fc4cdc

mullr commented on src/rsl.rs:
```diff
@@ -232,7 +237,8 @@ impl<'remote, 'repo> RSL<'remote, 'repo> {
         let mut current = Some(self.remote_head);
-        let ref_name = format!("refs/heads/{}", branch);
+        let reference = git::get_ref_from_name(self.repo, branch).expect("failed to get ref");
+        let  ref_name = reference.name().expect("failed to get ref name");
```
extra space after let

clearydude replied:
2fc4cdc

windwardrail commented on src/bin/git-secure-fetch.rs:
```diff
@@ -32,8 +32,8 @@ fn main() {
-    let branch = match matches.value_of("BRANCH") {
-        None => panic!("Must supply a BRANCH argument"),
+    let reference = match matches.value_of("REFERENCE") {
+        None => panic!("Must supply a REFERENCE argument"),
```
It would be nice to avoid panics. You could use `bail!` from error chain if you moved the logic in main to another function and then called `handle_error` from main?

clearydude replied:
I think that this specific case should probably go into the issue tracker and be addressed as part of a different PR, since this one is really just supposed to be adding tag support.

windwardrail commented on src/push_entry.rs:
```diff
@@ -53,15 +53,12 @@ impl PushEntry {
-            .get()
-            .target()
-            .unwrap();
+        let full_ref = utils::git::get_ref_from_name(repo, branch_str).expect("failed to get reference");
```
Use error handling here instead of unwrapping with expect (cause it's basically just a panic)

clearydude replied:
638721c

windwardrail commented on src/push_entry.rs:
```diff
@@ -53,15 +53,12 @@ impl PushEntry {
-            .target()
-            .unwrap();
+        let full_ref = utils::git::get_ref_from_name(repo, branch_str).expect("failed to get reference");
+        let target_oid = full_ref.target().expect("failed to get target oid");
```
Error handling

clearydude replied:
638721c

windwardrail commented on src/push_entry.rs:
```diff
@@ -53,15 +53,12 @@ impl PushEntry {
         PushEntry {
-            ref_name: format!("refs/heads/{}", branch_str),
-            oid: branch_head,
+            ref_name: String::from(full_ref.name().expect("failed to get ref's full name")),
```
Error handling

clearydude replied:
638721c

windwardrail commented on src/rsl.rs:
```diff
@@ -225,14 +230,15 @@ impl<'remote, 'repo> RSL<'remote, 'repo> {
         let mut current = Some(self.remote_head);
-        let ref_name = format!("refs/heads/{}", branch);
+        let reference = git::get_ref_from_name(self.repo, reference).expect("failed to get ref");
+        let  ref_name = reference.name().expect("failed to get ref name");
```
Error handling

clearydude replied:
638721c

windwardrail commented on tests/integration_test.rs:
```diff
@@ -30,36 +30,54 @@ macro_rules! sequential_test {
 
             do_work_on_branch(&context.local, "refs/heads/master");
-            let res4 = git_rsl::secure_push_with_cleanup(&mut context.local, &RemoteName::new("origin"), &BranchName::new("master")).expect("Could not run third push");
+            let res4 = git_rsl::secure_push_with_cleanup(&mut context.local, &RemoteName::new("origin"), &ReferenceName::new("master")).expect("Could not run third push");
```
These lines need to be wrapped somehow

clearydude replied:
2fc4cdc dig the fancy macro

windwardrail commented on tests/integration_test.rs:
```diff
@@ -30,36 +30,54 @@ macro_rules! sequential_test {
+            let remote_tag = &context.remote.find_reference("refs/tags/v6.66").expect("reference not found");
+            assert!(remote_tag.is_tag());
+
+            assert_eq!((), git_rsl::secure_fetch_with_cleanup(&mut context.local, &RemoteName::new("origin"), &ReferenceName::new("v6.66")).expect("could not fetch tag"));
```
Wrap lines

clearydude replied:
2fc4cdc

windwardrail commented on tests/integration_test.rs:
```diff
@@ -100,17 +118,17 @@ sequential_test! {
 
             do_work_on_branch(&context.local, "refs/heads/master");
-            let res3 = git_rsl::secure_push_with_cleanup(&mut context.local, &RemoteName::new("origin"), &BranchName::new("master")).expect_err("Checking for invalid RSL detection");
+            let res3 = git_rsl::secure_push_with_cleanup(&mut context.local, &RemoteName::new("origin"), &ReferenceName::new("master")).expect_err("Checking for invalid RSL detection");
```
Wrap long lines

clearydude replied:
2fc4cdc

windwardrail reviewed (changes requested):
Most of my comments are on error handling. I didn't flag all the instances, but using `expect` is essentially a `panic!` which we should avoid in favor of error handling such as Result or error chain.

Also there are long lines that should be wrapped. Didn't flag all of those either.

windwardrail commented:
Compiler warnings...
```
warning: unused import: `BranchType`
 --> src/push_entry.rs:5:18
  |
5 | use git2::{self, BranchType, Oid, Reference, Repository};
  |                  ^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default

warning: unused import: `Object`
  --> src/utils/test_helper.rs:13:24
   |
13 | use git2::{Repository, Object, ObjectType, Reference};
   |                        ^^^^^^

warning: unused variable: `e`
   --> src/utils/git.rs:273:13
    |
273 |         Err(e) => { // toDo may be other errors here...
    |             ^ help: consider using `_e` instead
    |
    = note: #[warn(unused_variables)] on by default

warning: unused variable: `tag_oid`
  --> src/utils/test_helper.rs:93:9
   |
93 |     let tag_oid = repo.tag_lightweight(tag_name, tag_target, false);
   |         ^^^^^^^ help: consider using `_tag_oid` instead
```

Commit:
Code cleanup

This commit cleans up some formatting issues and implements a macro for checking that method calls are ok in the integration tests file.

clearydude commented:
@windwardrail compiler warnings fixed as of 2fc4cdc

Commit:
Change to annotated tag

Prior to this commit, the tag test created a lightweight tag for tagging a specific release. This commit changes the helper method and test to instead create a more informative annotated tag.

Commit:
Use map for converting refnames

Prior to this commit, I was doing this awkwardly.
This commit uses map instead.

Commit:
Remove toDo's

Prior to this commit, the src files for the binaries still had toDo's in them.
Those toDos are still things that we need toDo, but would be better served being represented as issues in the issue tracker, so this commit removes them from the code.

Commit:
Add error handling to lib functions

Prior  to this commit, there were lots of panics in the new lib functions. This commit adds in error handling for those cases.

clearydude commented:
Addressed all the comments. I'd appreciate another quick look through when you have the chance @windwardrail @mullr
